src/main/java/com/icthh/xm/ms/entity/repository/search/elasticsearch/XmEntityMultipleIndexElasticIndexRepository.java

```diff
@@ -0,0 +1,121 @@
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+import org.springframework.transaction.annotation.Transactional;
+
+@ConditionalOnProperty(value = "application.use-elasticsearch-multiple-indices")
+@Component
+@RequiredArgsConstructor
+@Slf4j
+public class XmEntityMultipleIndexElasticIndexRepository implements XmEntityElasticRepository {
+
+    private final ElasticsearchTemplate elasticsearchTemplate;
+    private final MappingConfiguration mappingConfiguration;
+    private final IndexConfiguration indexConfiguration;
+    private final XmEntitySpecService xmEntitySpecService;
+    private final ElasticIndexNameResolver elasticIndexNameResolver;
+
+    private static final String XM_ENTITY_FIELD_TYPEKEY = "typeKey";
+
+    @Override
+    public XmEntity save(XmEntity xmEntity) {
+        IndexQuery fooIdxQuery = new IndexQueryBuilder()
+            .withIndexName(elasticIndexNameResolver.resolve(xmEntity.getTypeKey()))
+            .withObject(xmEntity)
+            .build();
+        elasticsearchTemplate.index(fooIdxQuery);
+        return xmEntity;
+    }
+
+    @Override
+    public void saveAll(List<XmEntity> xmEntity) {
+        List<IndexQuery> queries = xmEntity.stream()
+            .map(entity -> new IndexQueryBuilder()
+                .withIndexName(elasticIndexNameResolver.resolve(entity.getTypeKey()))
+                .withObject(entity)
+                .build())
+            .collect(Collectors.toList());
+        elasticsearchTemplate.bulkIndex(queries);
+    }
+
+    @Override
+    public void delete(XmEntity xmEntity) {
+        String typeKey = xmEntity.getTypeKey();
+        DeleteQuery deleteQuery = new DeleteQuery();
+        deleteQuery.setIndex(typeKey);
+        deleteQuery.setQuery(termQuery("id", xmEntity.getId()));
+        elasticsearchTemplate.delete(deleteQuery, XmEntity.class);
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public long handleReindex(Function<Specification, Long> specificationFunction) {
+        AtomicLong reIndexCount = new AtomicLong(0L);
+        List<TypeSpec> allTypes = xmEntitySpecService.findAllTypes();
+        List<String> typeKeyList = allTypes.stream().map(spec -> spec.getKey()).collect(Collectors.toList());
+        IntStream.iterate(0, i -> i < typeKeyList.size(), i -> i + 1).forEach(i -> {
+            String typeKey = typeKeyList.get(i);
+            log.info("{} from {} resolvedTypeKey [{}] is being indexed", i, typeKeyList.size(), typeKey);
+            Specification spec = (Specification) (root, query, criteriaBuilder) -> {
+                query.orderBy(criteriaBuilder.desc(root.get("id")));
+                return criteriaBuilder.equal(root.get(XM_ENTITY_FIELD_TYPEKEY), typeKey);
+            };
+            recreateIndex(typeKey);
+            Long count = specificationFunction.apply(spec);
+            reIndexCount.addAndGet(count);
+            if (!mappingConfiguration.isMappingExists()) {
```

Comment:
mappingConfiguration need to be improved for keep separate configuration for each index

Reply:
i think it worth to keep some global(default) mapping file and add functionality to replace it with mapping per index.